### PR TITLE
Flink: improve metrics (elapsedSecondsSinceLastSuccessfulCommit) and …

### PR DESCRIPTION
**Pull request by stevenzwu**
…logging for IcebergFilesCommitter

### CODE REVIEW

File: flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java [hunk 1]
```diff
@@ -230,6 +230,7 @@ public void notifyCheckpointComplete(long checkpointId) throws Exception {
     // the files,
     // Besides, we need to maintain the max-committed-checkpoint-id to be increasing.
     if (checkpointId > maxCommittedCheckpointId) {
+      LOG.info("Checkpoint {} completed. Attempting commit.", checkpointId);
```
Comment (stevenzwu) [inline flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:233]: add those INFO logging for easier navigation of code paths

File: flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitterMetrics.java [hunk 2]
```diff
@@ -37,6 +40,9 @@ class IcebergFilesCommitterMetrics {
         metrics.addGroup("IcebergFilesCommitter").addGroup("table", fullTableName);
     committerMetrics.gauge("lastCheckpointDurationMs", lastCheckpointDurationMs::get);
     committerMetrics.gauge("lastCommitDurationMs", lastCommitDurationMs::get);
+    this.elapsedSecondsSinceLastSuccessfulCommit = new ElapsedTimeGauge(TimeUnit.SECONDS);
```
Comment (stevenzwu) [inline flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitterMetrics.java:43]: Iceberg commit happens after a successful checkpoint. Iceberg commits can fail (for whatever reason) while Flink checkpoints succeeded. this metric is very useful to set up alerts to detect consecutive Iceberg commit failures.

Will send a separate PR for doc on FlinkSink metrics.

**Review by pvary — APPROVED** [review]

File: flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitterMetrics.java (hunk 2 above)
Comment (stevenzwu) [reply flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitterMetrics.java:43]: PR #6765 for doc update

File: flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitterMetrics.java [hunk 3]
```diff
@@ -54,12 +60,34 @@ void commitDuration(long commitDurationMs) {
...
+      this.lastSuccessfulCommitTimeNano = System.nanoTime();
+    }
+
+    void refreshLastSuccessfulCommitTime() {
+      this.lastSuccessfulCommitTimeNano = System.nanoTime();
+    }
+
+    @Override
+    public Long getValue() {
+      return reportUnit.convert(
```
Comment (mas-chen) [inline flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitterMetrics.java:89]: It's possible that this metric may polled by the metric reporter before a single checkpoint has completed. Can that edge case be handled? Reporting a huge number is going to be confusing

Comment (stevenzwu) [reply flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitterMetrics.java:89]: > It's possible that this metric may polled by the metric reporter before a single checkpoint has completed. 

@mas-chen you were referring to the first checkpoint, right? this is not an issue. Note that the report value is a delta value as `System.nanoTime() - lastSuccessfulCommitTimeNano`.

Comment (mas-chen) [reply flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitterMetrics.java:89]: No, before the first checkpoint. `lastSuccessfulCommitTimeNano` would be 0 and the delta would be large

Comment (mas-chen) [reply flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitterMetrics.java:89]: I missed the constructor init for that variable. Nevermind!

File: flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitterMetrics.java [hunk 4]
```diff
@@ -54,12 +60,38 @@ void commitDuration(long commitDurationMs) {
...
+  private static class ElapsedTimeGauge implements Gauge<Long> {
+    private final TimeUnit reportUnit;
+    private volatile long lastRecordedTimeNano;
+
+    ElapsedTimeGauge(TimeUnit timeUnit) {
+      this.reportUnit = timeUnit;
+      this.lastRecordedTimeNano = System.nanoTime();
+    }
+
+    void refreshLastRecordedTime() {
```
Comment (stevenzwu) [inline flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitterMetrics.java:87]: @pvary I renamed this method from `refreshLastSuccessfulCommitTime` to `refreshLastRecordedTime`, as the class `ElapsedTimeGauge` was intended as a more general gauge.

Comment (pvary) [reply flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitterMetrics.java:87]: +1

**Pull request merged**